staging/src/k8s.io/cloud-provider/controllers/nodelifecycle/node_lifecycle_controller.go

```diff
@@ -196,18 +196,49 @@ func (c *CloudNodeLifecycleController) MonitorNodes(ctx context.Context) {
 	}
 }
 
+// getProviderID returns the provider ID for the node. If Node CR has no provider ID,
+// it will be the one from the cloud provider.
+func (c *CloudNodeLifecycleController) getProviderID(ctx context.Context, node *v1.Node) (string, error) {
+	if node.Spec.ProviderID != "" {
+		return node.Spec.ProviderID, nil
+	}
+
+	if instanceV2, ok := c.cloud.InstancesV2(); ok {
+		metadata, err := instanceV2.InstanceMetadata(ctx, node)
+		if err != nil {
+			if err == cloudprovider.InstanceNotFound {
+				return "", nil
+			}
+			return "", err
+		}
+		return metadata.ProviderID, nil
+	}
+
+	providerID, err := cloudprovider.GetInstanceProviderID(ctx, c.cloud, types.NodeName(node.Name))
```

Comment:
With your implementation we cannot return the correct value if `cloudprovider.InstanceNotFound`.
Please take a look at my implementation in PR https://github.com/kubernetes/kubernetes/pull/121383, I had considered everything there IMO

Reply:
Yeah I noticed this, I kind of ignored this because I feel this will be rewritten anyway (look at my bigger comment below) but I will definitely keep that in mind to handle `cloudprovider.InstanceNotFound` once we agree on the high-level shape of things